newrelic-scala-api/src/main/scala/com/newrelic/scala/api/TraceOps.scala

```diff
@@ -0,0 +1,164 @@
+   * @param f           Function segment is to capture metrics for.
+   * @tparam T Input type for function segment is to capture metrics for.
+   * @tparam S Type returned from executed function
+   * @return Value returned from executed function
+   */
+  def traceFun[T, S](segmentName: String)(f: T => S): T => S = {
+    t: T =>
+      val txn = NewRelic.getAgent.getTransaction()
+      val segment = txn.startSegment(segmentName)
+      try {
+        f(t)
+      } finally {
+        segment.end()
+      }
+  }
+
+  /**
+   * Creates a segment to capture metrics for a given asynchronous function, this will call {@link com.newrelic.api.agent.Transaction#startSegment(String)},
+   * execute the function, then call {@link com.newrelic.api.agent.Segment#end()} on the <b>completion</b> of the asynchronous function.
+   * This {@link Segment} will show up in the Transaction Breakdown table, as well as the Transaction Trace page. This {@link Segment} will be reported in the "Custom/" metric
+   * e.g. The code below will produce 1 segment <i>trace flatMap segment</i>
+   * <pre>
+   * Future(1).flatMap(asyncTraceFun("trace flatMap segment")(i => Future(i + 1)))
+   * </pre>
+   *
+   * @param segmentName Name of the { @link Segment} segment in APM.
+   *                    This name will show up in the Transaction Breakdown table, as well as the Transaction Trace page.
+   *                    <p>
+   *                    if null or an empty String, the agent will report "Unnamed Segment".
+   * @param f Asynchronous function segment is to capture metrics for.
+   * @param ec The execution context on which the future is run
+   * @tparam T Input type for function segment is to capture metrics for.
+   * @tparam S Type returned from completed asynchronous function
+   * @return Value returned from completed asynchronous function
+   */
+  def asyncTraceFun[T, S](segmentName: String)(f: T => Future[S])(implicit ec: ExecutionContext): T => Future[S] = {
+    t: T =>
+      val txn = NewRelic.getAgent.getTransaction()
+      val segment = txn.startSegment(segmentName)
+      val evaluatedFunc = f(t)
+      evaluatedFunc.onComplete {
+        case _ => segment.end()
+      }
+      evaluatedFunc
+
+  }
+
+  /**
+   * Starts a {@link com.newrelic.api.agent.Transaction} for a given block of code.
+   * When this method is invoked within the context of an existing transaction this has no effect.
+   * The newly created {@link com.newrelic.api.agent.Transaction} will complete once the code block has been executed
+   * e.g. the code below will create a Transaction and with 2 segments <i>trace option creation</i> and <i>trace map option</i>
+   * <pre>
+   * txn {
+   *    val o1 = trace("trace option creation")(Some(1))
+   *    o1.map(traceFun("trace map segment")(i => i + 1))
+   * }
+   * </pre>
+   * e.g. the code below will create a Transaction and with 2 segments <i>trace future creation</i> and <i>trace filter future</i>
```

Comment:
Again, maybe I'm missing something, but the names don't seem to match what is actually in the example.

Given the code same below, shouldn't this be

```
<i>trace map segment</i> and <i>trace filter segment</i>
```

instead of

```
<i>trace future creation</i> and <i>trace filter future</i>
```
?

Reply:
changes added to PR